### ISSUE TITLE
Support pypi /simple endpoints not ending on /simple 

### ISSUE THREAD

**Issue by hesstobi** [renovatebot/renovate#3864]
**What would you like Renovate to be able to do?**
I have a private pypi index server which servers only the simple api. But the endpoint is not ending on /simple. Thus renovate currently try the json api on that index server, but this will of course fail. It would be nice if renovate will support also the simple api on such endpoints.

**Describe the solution you'd like**
https://github.com/renovatebot/renovate/blob/11befb119c01e7b3d2f3418afa824dd4ce4c565d/lib/datasource/pypi/index.js#L41

To switch betwenn the simple and the json api renovat just look if the `hostUrl` ends with `simple`. It would be more flexible if the type of the api is checked based on the response from the host. 

**Describe alternatives you've considered**
An alternative solution may be, to add option with list of  urls which should be conconsidered having the simple api.

**Comment by nihal9ns** [comment]
@hesstobi Can you please provide an example url? Your private pypi index server url will also help. If you can give a list of urls which should be considered would be even more helpful.

**Comment by nihal9ns** [comment]
@hesstobi Also, I wanted to know if this doesn't work for you? 
```if (hostUrl.endsWith('/simple/')) {
      dep = await getSimpleDependency(lookupName, hostUrl);
    } else {
      dep = await getDependency(lookupName, hostUrl, compatibility);
    }```

**Comment by rarkins** [comment]
I think the issue is that he has a server like `https://some.private.server/foo` which can't be detected based on URL suffix alone. My question is: what approach can Renovate use to determine that this is a simple endpoint? e.g. can we have more details for this part?

> the type of the api is checked based on the response from the host

**Comment by hesstobi** [comment]
My servers having urls like `https://example.com/repo/default` and `https://example.com/repo/other`. 

It seems there is no standard way to check if there url is providing a json or a simple api. Would be a solution to remove the if/else and request both types and return the one which is found.

```js

let [debSimple, debJson] = await Promise.all([getSimpleDependency(lookupName, hostUrl), getDependency(lookupName, hostUrl, compatibility)]);

if (debSimple !== null) {
      return debSimple;
}

if (debJson !== null) {
      return debJson;
}

```

Of course the console warnings should then be move from the `getDependency` function to the `getPkgReleases`

**Comment by rarkins** [comment]
This sounds like a reasonable approach. I'd prefer to not have "intentionally failing" requests but unless we think of something else, this might be the best. We could for example try simple first and then complex, but then for registries that support both we'd end up with decreased metadata.

**Comment by hesstobi** [comment]
Should I try to prepare a MR for this? 
But I would try the JSON API first to avoid decreased metadata.

**Comment by rarkins** [comment]
I think it would be best if we can work out how `pip` does this. e.g. does it try all in parallel, or try/fail/try, etc?

**Comment by hesstobi** [comment]
I did a closer look how pip is doing it. If I understand it correctly it just uses the simple API for listing packages see https://github.com/pypa/pip/blob/master/docs/html/development/architecture/package-finding.rst.

**Comment by RmStorm** [comment]
> I did a closer look how pip is doing it. If I understand it correctly it just uses the simple API for listing packages see https://github.com/pypa/pip/blob/master/docs/html/development/architecture/package-finding.rst.

I think you're right, it seems that pip relies exclusively on the simple API for downloading installing and listing packages.

> My servers having urls like `https://example.com/repo/default` and `https://example.com/repo/other`.
> 
> It seems there is no standard way to check if there url is providing a json or a simple api. Would be a solution to remove the if/else and request both types and return the one which is found.
> 
> ```js
> let [debSimple, debJson] = await Promise.all([getSimpleDependency(lookupName, hostUrl), getDependency(lookupName, hostUrl, compatibility)]);
> 
> if (debSimple !== null) {
>       return debSimple;
> }
> 
> if (debJson !== null) {
>       return debJson;
> }
> ```
> 
> Of course the console warnings should then be move from the `getDependency` function to the `getPkgReleases`

If @rarkins agrees I will implement this snippet and make a PR? I do think debSimple and debJson should have their priorities reversed:
```js
let [debSimple, debJson] = await Promise.all([getSimpleDependency(lookupName, hostUrl), getDependency(lookupName, hostUrl, compatibility)]);

if (debJson !== null) {
      return debJson;
}

if (debSimple !== null) {
      return debSimple;
}

if (debSimple !== null) && (debJson !== null) {
      return null;
}
```
e.g. json should be prioritised if available, not the other way around.

**Comment by rarkins** [comment]
If pip always uses simple endpoints then that's an interesting discovery (thanks). I was not aware that every pypi server supports simple - I thought they had the choice of supporting a proper API, simple only, or both.

If that's the case then I think then what we want to do is:
- always use simple to get the list of releases
- use APIs as a second strategy for enriching the package/releases with metadata (such as source URLs, release dates, etc). We can cache that data pretty aggressively, e.g. only requery it every 24 hours.

**Comment by RmStorm** [comment]
Yes, I did some additional testing and it really seems like pip treats everything as a [simple api](https://www.python.org/dev/peps/pep-0503/). However it does so in a not so very clever way.

If you supply it with ``--index-url https://pypi.org/pypi`` It will look up the url ``https://pypi.org/pypi/packagename`` instead of understanding it points to a json api and using ``https://pypi.org/pypi/packagename/json``. So what happens next is that pypi redirects to ``https://pypi.org/project/packagename/`` for example: https://pypi.org/project/Django/

And then pip scans that whole page reading all links and checking if any of the links are compliant with PEP503! and if it finds any it will use those. So yeah, currently pip completely disregards the json api but it doesn't actually try to find simple endpoints either. It just treats every endpoint as a simple endpoint.

**Comment by rarkins** [comment]
Let's not emulate that behaviour :D

So would this make sense?

1. Always try the simple endpoint protocol first. If it succeeds, then also check for JSON endpoint to enrich the package and release metadata. Use a long cache time for this the JSON metadata.
2. If simple fails (not sure we have a use case for that?) then fall back to the JSON API with regular cache times (e.g. 10 minutes)

**Comment by RmStorm** [comment]
yes, so the thing is trying the simple endpoint protocol on any hostUrl is basically emulating the pip behaviour and I would wager it will rarely if ever fail. So I don't think you would ever hit case 2 but we can program it in that way just in case.

**Comment by rarkins** [comment]
Now to think through the logistics.

Case 1: A url like `https://some.server/simple/`. In this case that's the actual URL we use, so it's easy. This one already works today.

Case 2: A url like `https://another.server/foo/`. In this case it's a simple endpoint but /simple/ isn't in the URL. We don't handle this today because we don't detect it as simple. I guess though in the new implementation we just try it first as simple and succeed.

Case 3: `https:/pypi.org/simple/`. This works

Case 4: `https://pypi.org`. This URL does not work directly, because it's a regular webpage. So we try it and fail. Do we then always add `simple/` to the URL and try again?

Case 5: `https://pypi.org/pypi`. It redirects to `https://pypi.org`. Are we expected to follow that redirect, fail, and *then* add `/simple` to the end of the final url after redirect? It does not work to add `simple/` to the original URL because `pypi.org/pypi/simple` redirects to https://pypi.org/project/simple/

**Comment by rarkins** [comment]
Cases 1-3 above all just work if we test the URL we're given for simple. Case 4 only works if we add simple/ to the URL we're given. Case 5 only works if we follow a redirect and *then* add simple/. Is Case 5 something we need to support?

**Comment by RmStorm** [comment]
Case 4 is not supported by pip so I don't think we should support that. Case 5 is the big question I guess. Currently trying the simple protocol on case 5 would actually work as far as I can tell.
https://github.com/renovatebot/renovate/blob/11befb119c01e7b3d2f3418afa824dd4ce4c565d/lib/datasource/pypi/index.js#L103
We would basically look up ```https://pypi.org/pypi/depName``` which redirects to ```https://pypi.org/project/depName```. That page will have valid simple links as checked by:
https://github.com/renovatebot/renovate/blob/11befb119c01e7b3d2f3418afa824dd4ce4c565d/lib/datasource/pypi/index.js#L139-L145

**Comment by rarkins** [comment]
When you say valid links, I was expecting something like on https://pypi.org/simple/acalib/ and not like on https://pypi.org/project/acalib/

I see only "non simple" html like this:

```
              <a class="card release__card" href="/project/acalib/0.1.2/">
                <p class="release__version">
                  0.1.2 
                </p>
                <p class="release__version-date">
                  <time class="tooltipped tooltipped-s -js-relative-time" datetime="2017-09-15T03:52:27+0000" data-controller="localized-time">
  Sep 15, 2017
</time>
                </p>
              </a>
```

**Comment by RmStorm** [comment]
By valid link I mean a link that conforms to the standard in [PEP 503](https://www.python.org/dev/peps/pep-0503/). So I don't mean a hostUrl that points to a simple or a json API. I mean a link that refers to a specific version of a specific package.

The piece of code in pip that evaluates a link for validity seems to be this:
https://github.com/pypa/pip/blob/master/src/pip/_internal/index.py#L408-L494

Currently the getSimpleDependency also loops through all links in a webpage and tries to find valid links using a single regexp. If you run getSimpleDependency on the url for a json API (basically case5) you run into trouble. Just the same if case2 is actually a json endpoint instead of a simple endpoint.

So what I mean is that if we want to run getSimpleDependency on url's that might not actually be a simple endpoint we might run into a lot of false positives using ``const versionRegexp = /\d+(\.\d+)+/;``. I think there might be many links that match without actually being valid links.

**Comment by RmStorm** [comment]
> When you say valid links, I was expecting something like on https://pypi.org/simple/acalib/ and not like on https://pypi.org/project/acalib/
> 
> I see only "non simple" html like this:
> 
> ```
>               <a class="card release__card" href="/project/acalib/0.1.2/">
>                 <p class="release__version">
>                   0.1.2 
>                 </p>
>                 <p class="release__version-date">
>                   <time class="tooltipped tooltipped-s -js-relative-time" datetime="2017-09-15T03:52:27+0000" data-controller="localized-time">
>   Sep 15, 2017
> </time>
>                 </p>
>               </a>
> ```

In this webpage: https://pypi.org/project/acalib/
``view-source:https://pypi.org/project/acalib/``
Line 722-724 and 743-745 would constitute valid links.

ps.: I have no idea if this works? does ``view-source:`` work on every browser or is it firefox specific? also if there is some dynamically injected html the line numbers won't work out...

**Comment by RmStorm** [comment]
I think the conclusion is that if you want to support simple API's that are not marked as such by having their url ending on ``/simple/`` there are two options:

1. you need to have some logic to determine if a page is actually a simple api endpoint
2. you need to have logic to parse every link on a page in the same way that pip does

One of these two options has to be implemented to resolve this issue.

**Comment by RmStorm** [comment]
> Let's not emulate that behaviour :D
> 
> So would this make sense?
> 
>     1. Always try the simple endpoint protocol first. If it succeeds, then also check for JSON endpoint to enrich the package and release metadata. Use a long cache time for this the JSON metadata.
> 
>     2. If simple fails (not sure we have a use case for that?) then fall back to the JSON API with regular cache times (e.g. 10 minutes)

Hi @rarkins I have some time today to implement this. Can you give me an example of how to change the caching times? I tried finding an example in the renovatebot codebase but I haven't figured it out. Also how long do you think the caching times should be on the json metadata when it is used to supplement a simple endpoint? 24 hours?

**Comment by rarkins** [comment]
@RmStorm if you just implement the logic, then I'm happy to fix up the caching afterwards if you like. I only just realised we maybe have *no* caching on PyPi right now anyway. If you would like to see an example of caching, search "cache" in this file: https://github.com/renovatebot/renovate/blob/master/lib/datasource/nuget/v3.ts

Focus on the logic first though, I'm happy to fill in the cache details!

**Comment by hesstobi** [comment]
Is there any progress on this issue? Can I do anything to get the fixed soon?

**Comment by rarkins** [comment]
@hesstobi you're welcome to take up where @RmStorm left off. i.e. implementing the logic as last discussed.

**Comment by RmStorm** [comment]
Hi @hesstobi, sadly I haven't yet found the time to work on this. I don't think I will have time in the near future either.

Most of the our dependencies are on pypi so the few dependencies that we have to update by hand are not that big a problem for us. I would still like to have it fixed off course but it's quite low on my todo list for these reasons.

**Comment by TheKevJames** [comment]
FYI: I created #6649 before seeing the conversation here, which adds a "quick fix" for this issue -- if the json api does not exist, try the simple api before giving up.

**Comment by rarkins** [comment]
#6649 should solve this problem but I would like to keep this issue open for an improved solution. The downside of #6649 is that it will generate one 404 error per dependency, whereas ideally we can do a single "check" of a registry if it's simple or not and then remember that.
